File: ProblemSolve/1046.cpp
/*********************
 * Last Stone Weight *
 *********************/

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> pq;
        
        for(int stone : stones){
            pq.push(stone);
        }
        
        while(pq.size() > 1){
            int stoneY = pq.top();
            pq.pop();
            int stoneX = pq.top();
            pq.pop();
            
            if(stoneX < stoneY){
                pq.push(stoneY - stoneX);
            }
        }
        
        return pq.size() == 1 ? pq.top() : 0;
    }
};

File: ProblemSolve/1046.cs
/*********************
 * Last Stone Weight *
 *********************/

public class Solution {
    public int LastStoneWeight(int[] stones) {
        PriorityQueue<int, int> pq = new PriorityQueue<int, int>();
        
        foreach(int stone in stones){
            pq.Enqueue(stone, -stone);
        }
        
        while(pq.Count > 1){
            int stoneY = pq.Dequeue();
            int stoneX = pq.Dequeue();
            
            Console.WriteLine(stoneY);
            Console.WriteLine(stoneX);
            
            if(stoneX < stoneY){
                pq.Enqueue(stoneY - stoneX, stoneX - stoneY);
            }
        }
        
        return pq.Count == 1 ? pq.Dequeue() : 0;
    }
}